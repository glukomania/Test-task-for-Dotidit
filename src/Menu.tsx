import * as React from 'react'

type Header = {
  Header: string
  accessor: string
  isVisible: boolean
}

type MenuProps = {
  headers: Header[]
  setHeaders: (newHeaders: Header[]) => void;
  header: string
  setIsMenuOpen: (arg0: boolean) => void;
  isMenuOpen: boolean
}

const Menu = (props: MenuProps) => {

  const onColumnToggle = (name: string) => {

    if (props.headers.filter(item => item.isVisible).length > 1) {        //logic to prevent the whole table disappearing
      const selectedHeader = props.headers.find((header: Header) => header.Header === name)
      const newSelectedHeader = {...selectedHeader, isVisible: !selectedHeader.isVisible}
  
      const newHeaders = props.headers.map((header: Header) => (header.Header === newSelectedHeader.Header ? newSelectedHeader : header));
  
      props.setHeaders(newHeaders)
      
    } else {
      const selectedHeader = props.headers.find((header: Header) => header.Header === name)
      if (!selectedHeader.isVisible) {
        const newSelectedHeader = {...selectedHeader, isVisible: !selectedHeader.isVisible}
  
        const newHeaders = props.headers.map((header: Header) => (header.Header === newSelectedHeader.Header ? newSelectedHeader : header));
    
        props.setHeaders(newHeaders)
      }
    }
  }

  return (
    <div className="menu">
      <div className="blur" onClick={() => props.setIsMenuOpen(!props.isMenuOpen)}/>
      <div className="menu__content">
        <div className='menu__header'>{props.header}</div>
          <ul>
            {props.headers.map((item: Header) => {
              return (<div key={item.Header} className="menuitem">
                        <label>
                          <input
                            type="checkbox"
                            checked={item.isVisible}
                            onChange={() => onColumnToggle(item.Header)}
                            className='checkbox'
                          />
                          {item.Header}
                        </label>
                      </div>)}
            )}
          </ul>
      </div>
    </div>
  )
}

export default Menu