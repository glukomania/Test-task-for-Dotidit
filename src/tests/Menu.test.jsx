import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import Menu from '../Menu';

const headersMock = [
  { Header: 'Column1', accessor: 'col1', isVisible: true },
  { Header: 'Column2', accessor: 'col2', isVisible: false },
  { Header: 'Column3', accessor: 'col3', isVisible: true },
];

describe('Menu component', () => {
  it('renders without crashing', () => {
    render(
      <Menu
        headers={headersMock}
        setHeaders={() => {}}
        header="Test Header"
        setIsMenuOpen={() => {}}
        isMenuOpen={false}
      />
    );
  });

  it('renders the correct number of checkboxes', () => {
    const { container } = render(
      <Menu
        headers={headersMock}
        setHeaders={() => {}}
        header="Test Header"
        setIsMenuOpen={() => {}}
        isMenuOpen={false}
      />
    );

    const checkboxes = container.querySelectorAll('input[type="checkbox"]');
    expect(checkboxes.length).toBe(headersMock.length);
  });

  it('toggles the visibility of a column on checkbox change', () => {
    const setHeadersMock = jest.fn();
    const { getByText } = render(
      <Menu
        headers={headersMock}
        setHeaders={setHeadersMock}
        header="Test Header"
        setIsMenuOpen={() => {}}
        isMenuOpen={false}
      />
    );

    fireEvent.click(getByText('Column2').closest('label'));

    expect(setHeadersMock).toHaveBeenCalledWith([
      { Header: 'Column1', accessor: 'col1', isVisible: true },
      { Header: 'Column2', accessor: 'col2', isVisible: true },
      { Header: 'Column3', accessor: 'col3', isVisible: true },
    ]);
  });


  it('closes the menu when the blur area is clicked', () => {
    const setIsMenuOpenMock = jest.fn();
    const { container } = render(      
    <Menu
      headers={headersMock}
      setHeaders={() => jest.fn()}
      header="Test Header"
      setIsMenuOpen={setIsMenuOpenMock}
      isMenuOpen={true}
    />);
    const blurArea = container.querySelector('.blur');
    fireEvent.click(blurArea);

    expect(setIsMenuOpenMock).toHaveBeenCalledWith(false);
  });
});