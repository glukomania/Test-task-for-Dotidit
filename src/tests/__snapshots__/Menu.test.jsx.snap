// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu component renders correctly 1`] = `
<div>
  <div
    class="menu"
  >
    <div
      class="blur"
    />
    <div
      class="menu__content"
    >
      <div
        class="menu__header"
      >
        Menu Header
      </div>
      <ul>
        <div
          class="menuitem"
        >
          <label>
            <input
              checked=""
              class="checkbox"
              disabled=""
              type="checkbox"
            />
            Name
          </label>
        </div>
        <div
          class="menuitem"
        >
          <label>
            <input
              checked=""
              class="checkbox"
              type="checkbox"
            />
            Age
          </label>
        </div>
        <div
          class="menuitem"
        >
          <label>
            <input
              class="checkbox"
              type="checkbox"
            />
            City
          </label>
        </div>
      </ul>
    </div>
  </div>
</div>
`;
